### PR TITLE
Upgrading to MasterChefV2

### PR DESCRIPTION
**Pull request by hexonaut**
API is fairly similar. Some small changes in how rewards are calculated. There is one more potentially dangerous operation added to V2 - the `set` function which updates the `rewarder` contract for the pool. This will only allow `cage()` to be called if `overwrite` is set to true.

Added some additional checks to confirm the dangerous operations do what they are supposed to.

Waiting for the `owner` to be updated before this is good to go.

### CODE REVIEW

File: src/sushi.sol [hunk 1]
```diff
@@ -186,19 +198,24 @@ contract SushiJoin is CropJoin {
         );
         require(
             selector == MasterChefLike.transferOwnership.selector ||
-            selector == MasterChefLike.setMigrator.selector
+            selector == MasterChefLike.setMigrator.selector ||
+            selector == MasterChefLike.set.selector
         , "SushiJoin/wrong-function");
+        if (selector == MasterChefLike.set.selector) {
+            uint8 overwrite = uint8(bytes(signature).length == 0 ? data[131] : data[127]) & 0x1;
```
Comment (kmbarry1) [inline src/sushi.sol:205]: Could we avoid the conditional here by reading byte 131 of the `callData` variable?

Comment (kmbarry1) [reply src/sushi.sol:205]: Also, more generally--is it really necessary to support passing an empty `signature`, since it seems that in that case the caller just has to include the selector in `data`? I personally don't see the need to support two ways of doing the same thing, and it makes the code more complex. Personally I'd remove the `signature` arg, rename `data` to `callData`, and then expect the first four bytes to be the selector (with appropriate checks).

File: src/sushi.sol [hunk 2]
```diff
@@ -186,19 +198,24 @@ contract SushiJoin is CropJoin {
         );
         require(
             selector == MasterChefLike.transferOwnership.selector ||
-            selector == MasterChefLike.setMigrator.selector
+            selector == MasterChefLike.setMigrator.selector ||
+            selector == MasterChefLike.set.selector
         , "SushiJoin/wrong-function");
+        if (selector == MasterChefLike.set.selector) {
```
Comment (kmbarry1) [inline src/sushi.sol:204]: For `set`, shouldn't we also confirm that the `_pid` argument matches the `pid` value stored in this contract? Otherwise, an update of _any_ rewarder allows caging _any_ SushiJoin instance.

Assuming I'm looking at the correct implementation:
https://github.com/sushiswap/sushiswap/blob/master/contracts/MasterChefV2.sol#L131

File: src/test/sushi-integration.t.sol [hunk 3]
```diff
@@ -675,19 +737,68 @@ contract SushiIntegrationTest is TestBase {
...
+            address(masterchef),
+            0,
+            "",
+            abi.encodeWithSelector(MasterChefLike.set.selector, join.pid(), 100, address(user2), true),
+            block.timestamp
+        );
+        assertEq(masterchef.rewarder(join.pid()), address(user2));
     }
 
     function testFail_cage_queued_irrelevant_change() public {
```
Comment (kmbarry1) [inline src/test/sushi-integration.t.sol:804]: Should have another test like this cover the "different pool id" irrelevant scenario.

File: src/sushi.sol [hunk 4]
```diff
@@ -58,6 +63,7 @@ contract SushiJoin is CropJoin {
 
     MasterChefLike  immutable public masterchef;
     address                   public initialMigrator;
+    address                   public initialRewarder;
     TimelockLike              public timelockOwner;
     uint256                   public pid;
```
Comment (kmbarry1) [inline src/sushi.sol:68]: Looks like this can't change on the MC side--any reason it can't be `immutable`?

File: src/sushi.sol [hunk 5]
```diff
@@ -58,6 +63,7 @@ contract SushiJoin is CropJoin {
 
     MasterChefLike  immutable public masterchef;
     address                   public initialMigrator;
+    address                   public initialRewarder;
     TimelockLike              public timelockOwner;
```
Comment (kmbarry1) [inline src/sushi.sol:67]: I'm confused by the naming here--seems like this is the Timelock itself, so why is the variable called "timelockOwner"?

File: src/sushi.sol [hunk 6]
```diff
@@ -58,6 +63,7 @@ contract SushiJoin is CropJoin {
 
     MasterChefLike  immutable public masterchef;
     address                   public initialMigrator;
+    address                   public initialRewarder;
```
Comment (kmbarry1) [inline src/sushi.sol:66]: Since these can both be updated, I'd drop the prefix "initial". If you still want to be descriptive, you could prefix them with e.g. "allowed".

File: src/sushi.sol (hunk 1 above)
Comment (hexonaut) [reply src/sushi.sol:205]: Updated the `data` to be `callData`.  The double way to specify the signature and calldata is kind of dumb I agree, but I have to copy what is in the timelock contract. Removing one or the other would allow a valid signature to be specified in the timelock for which we could not cage on as `bytes32 txHash = keccak256(abi.encode(masterchef, value, signature, data, eta));` would change depending on the style used.

File: src/sushi.sol (hunk 4 above)
Comment (hexonaut) [reply src/sushi.sol:68]: This can be changed via the `set()` function on MC with the `overwrite` flag set to true. https://github.com/sushiswap/sushiswap/blob/master/contracts/MasterChefV2.sol#L134

File: src/sushi.sol (hunk 5 above)
Comment (hexonaut) [reply src/sushi.sol:67]: This is the `owner` of the MC contract although I agree the names are a little ugly. I'll make a change.

File: src/sushi.sol (hunk 4 above)
Comment (kmbarry1) [reply src/sushi.sol:70]: That line you've highlighted is changing the `rewarder`, not the `pid`. I still don't see where the `pid` can change--as it's always an index into the `poolInfo` array.

File: src/test/base.sol [hunk 7]
```diff
@@ -141,4 +141,36 @@ contract TestBase is DSTest {
...
+
+    function assertEqApproxBPS(uint256 _a, uint256 _b, uint256 _tolerance_bps) internal {
+        uint256 a = _a;
+        uint256 b = _b;
+        if (a < b) {
+            uint256 tmp = a;
+            a = b;
+            b = tmp;
+        }
+        if (a - b > mul(_b, _tolerance_bps) / 10 ** 5) {
```
Comment (kmbarry1) [inline src/test/base.sol:168]: For basis points you should use 10^4, not 10^5: https://en.wikipedia.org/wiki/Basis_point

File: src/test/sushi-integration.t.sol [hunk 8]
```diff
@@ -278,26 +321,26 @@ contract SushiIntegrationTest is TestBase {
...
+            assertEqApprox(usr.sushi(), psushi + sushiToUser, 1);
             if (join.total() > 0) {
-                assertEq(join.stock(), pstock + punclaimedRewards - sushiToUser);
+                assertEqApprox(join.stock(), pstock + punclaimedRewards - sushiToUser, 1);
             } else {
                 assertTrue(join.stock() <= dust);  // May be a slight rounding error
             }
             if (ptotal > 0) {
-                assertEq(join.share(), pshare + rdiv(punclaimedRewards, ptotal));
+                assertEqApproxBPS(join.share(), pshare + rdiv(punclaimedRewards, ptotal), 500);
```
Comment (kmbarry1) [inline src/test/sushi-integration.t.sol:331]: 500 bps is pretty big (5%), but as noted right now the `assertEqApproxBPS` is using 10^5 instead of 10^4, so this is currently a 0.5% tolerance; make sure to update this and other invocations once `assertEqApproxBPS` is fixed.

File: src/test/sushi-integration.t.sol [hunk 9]
```diff
@@ -627,93 +671,159 @@ contract SushiIntegrationTest is TestBase {
...
+
+    function test_cage_queued_safe_change_safe_args5() public {
+        queue_safe_timelock_action(address(masterchef), "set(uint256,uint256,address,bool)", abi.encode(join.pid() + 1, 100, address(user2), true), false);
+    }
+
+    function test_cage_queued_safe_change_safe_args6() public {
+        queue_safe_timelock_action(address(masterchef), "", abi.encodeWithSelector(MasterChefLike.set.selector, join.pid() + 1, 100, address(user2), true), false);
+    }
+
+    // Verify that we can't get around the hash comparison by providing unexpected bool values
```
Comment (kmbarry1) [inline src/test/sushi-integration.t.sol:798]: Might it be better to use fuzzing here instead, or in addition (if these specific cases are uniquely important)?

File: src/test/sushi-integration.t.sol [hunk 10]
```diff
@@ -627,93 +671,159 @@ contract SushiIntegrationTest is TestBase {
...
+        queue_safe_timelock_action(address(masterchef), "", abi.encodeWithSelector(MasterChefLike.set.selector, join.pid(), 100, address(rewarder), true), true);
+    }
+
+    function test_cage_queued_safe_change_safe_args5() public {
+        queue_safe_timelock_action(address(masterchef), "set(uint256,uint256,address,bool)", abi.encode(join.pid() + 1, 100, address(user2), true), false);
+    }
+
+    function test_cage_queued_safe_change_safe_args6() public {
+        queue_safe_timelock_action(address(masterchef), "", abi.encodeWithSelector(MasterChefLike.set.selector, join.pid() + 1, 100, address(user2), true), false);
+    }
```
Comment (kmbarry1) [inline src/test/sushi-integration.t.sol:796]: I love the thoroughness here, but could we go a step further and give the tests either more specific names or comments indicating why the change is "safe"?

Comment (kmbarry1) [reply src/test/sushi-integration.t.sol:796]: Also it seems in some invocations you're passing `execute` as `true` and in others, `false`. IIUC it could be false in all of these calls as all we care about is ensuring the adapter can't be `caged`. More generally I suspect you could factor+name the `queue`/`cage`/`execute` helpers more cleanly if you were so inclined--e.g. `queue_safe_timelock_action` is confusingly named because it also attempts to `cage` and optionally `executes`. I might take an approach like defining a `TimeLockTX` struct to make passing around the data easy, and then have simpler, more single-purpose functions like `queue_timelock_tx(TimeLockTX tx_info)`, `try_cage(TimeLockTX tx_info, bool expectSuccess)` and `execute_timelock_tx(TimeLockTX tx_info)`. Expressing the tests in terms of these might be more lines, but IMO would be far more readable and maintainable.

File: src/test/base.sol (hunk 7 above)
Comment (kmbarry1) [reply src/test/base.sol:168]: Also since you've sorted so that `a > b`, I'd use `b` on the RHS of the `>` instead of `_b` so that you're always using the tighter of the two possible error bounds.

Comment (kmbarry1) [reply src/test/base.sol:168]: Nvm the comment I just deleted if you saw it, since `_b` is interpreted as the expectation it's appropriate to use it as the basis for calculating the error margin.

File: src/sushi.sol [hunk 11]
```diff
@@ -129,29 +143,29 @@ contract SushiJoin is CropJoin {
 
     function crop() internal override returns (uint256) {
         if (live) {
-            // withdraw of 0 will give us only the rewards
-            masterchef.withdraw(pid, 0);
+            // This can possibly fail if no rewards are owed, but that doesn't really matter
```
Comment (kmbarry1) [inline src/sushi.sol:146]: nit: really ought to explain _why_ it doesn't matter

File: src/sushi.sol (hunk 1 above)
Comment (kmbarry1) [reply src/sushi.sol]: Ah yeah I see, we have no choice in the matter, the stupidity is forced upon us :upside_down_face: .

File: src/sushi.sol (hunk 4 above)
Comment (hexonaut) [reply src/sushi.sol:70]: Oops yes you're right. Immutable it is.

File: src/test/sushi-integration.t.sol (hunk 9 above)
Comment (hexonaut) [reply src/test/sushi-integration.t.sol:798]: Yeah these are uniquely important imo. Wanted to test two values < uint8 size but not 0 and 1 and the 1s bit set and not set along with the same test for > uint8 size. Anything other than 0 or 1 errors, so I'm satisfied.

File: src/test/sushi-integration.t.sol (hunk 10 above)
Comment (hexonaut) [reply src/test/sushi-integration.t.sol:796]: I'll add some comments to explain why things are safe and not safe. The reason I want to attempt to execute if available is because of this bug that almost made it through: https://github.com/makerdao/dss-crop-join/blob/master/src/sushi.sol#L33

As you can see that interface for `setMigrator` was the wrong interface. If I had execution on then the timelock execute would fail because it's an invalid call. It's just a double check the interfaces are correct.

Where I set the safe action to not execute, it's just an invalid call. Actually since more pools have been added `test_cage_queued_safe_change_safe_args5/6` is now valid, so I've turned those on. I get this could be prettied up a bit, but I think it works fine and I don't really want to spend too much time on it for cosmetic reasons. Maybe if I get some spare time.

File: src/sushi.sol [hunk 12]
```diff
@@ -85,23 +93,27 @@ contract SushiJoin is CropJoin {
         address bonus_,
         address masterchef_,
         uint256 pid_,
-        address initialMigrator_,
-        address timelockOwner_
+        address migrator_,
```
Comment (brianmcmichael) [inline src/sushi.sol:96]: `migrator_`, `rewarder_`, `timelock_`, and `bonus_`, and `gem_` constructor params are all unnecessary here, as they can be asserted via their public variables after deployment in test and before governance acceptance.

File: src/sushi.sol [hunk 13]
```diff
@@ -85,23 +93,27 @@ contract SushiJoin is CropJoin {
...
+        address rewarder_,
+        address timelock_
     )
         public
         CropJoin(vat_, ilk_, gem_, bonus_)
```
Comment (brianmcmichael) [inline src/sushi.sol:101]:
```suggestion
        CropJoin(vat_, ilk_, MasterChefLike(masterchef_).lpToken(pid_), MasterChefLike(masterchef_).SUSHI())
```

Then you can remove `bonus_` and `gem_` from the constructor params.

File: src/sushi.sol [hunk 14]
```diff
@@ -20,23 +20,30 @@ import "./crop.sol";
...
+    function emergencyWithdraw(uint256 _pid, address _to) external;
+    function transferOwnership(address _newOwner, bool _direct, bool _renounce) external;
+    function setMigrator(address) external;
+    function set(uint256 _pid, uint256 _allocPoint, address _rewarder, bool _overwrite) external;
+    function add(uint256 _allocPoint, address _lpToken, address _rewarder) external;
 }
 
 interface TimelockLike {
     function queuedTransactions(bytes32) external view returns (bool);
     function queueTransaction(address,uint256,string memory,bytes memory,uint256) external;
```
Comment (brianmcmichael) [inline src/sushi.sol:43]: nit, but function signatures here don't match that of `MasterChefLike`

File: src/sushi.sol [hunk 15]
```diff
@@ -110,14 +122,10 @@ contract SushiJoin is CropJoin {
...
-        if (what == "pid") pid = data;
-        else revert("SushiJoin/file-unrecognized-param");
-        emit File(what, data);
-    }
     function file(bytes32 what, address data) external auth {
-        if (what == "initialMigrator") initialMigrator = data;
-        else if (what == "timelockOwner") timelockOwner = TimelockLike(data);
+        if (what == "migrator") migrator = data;
+        else if (what == "rewarder") rewarder = data;
+        else if (what == "timelock") timelock = TimelockLike(data);
```
Comment (brianmcmichael) [inline src/sushi.sol:128]: Not sure how everyone feels about mixing `address` and interface values like this. I'd rather see these things defined as `address` and then invoked as `Interface(addr)` throughout, but I think we're inconsistent on it.

File: src/sushi.sol [hunk 16]
```diff
@@ -129,29 +137,29 @@ contract SushiJoin is CropJoin {
 
     function crop() internal override returns (uint256) {
         if (live) {
-            // withdraw of 0 will give us only the rewards
-            masterchef.withdraw(pid, 0);
+            // This can possibly fail if no rewards are owed, but that doesn't really matter as we wouldn't get any rewards anyways
+            try masterchef.harvest(pid, address(this)) {} catch {}
```
Comment (brianmcmichael) [inline src/sushi.sol:141]: We should have a test of this failure scenario so that we can ensure the passthrough works correctly here.

File: src/sushi.sol [hunk 17]
```diff
@@ -161,21 +169,28 @@ contract SushiJoin is CropJoin {
...
-    function cage(uint256 value, string memory signature, bytes memory data, uint256 eta) external {
+    function cage(uint256 value, string calldata signature, bytes calldata data, uint256 eta) external {
         require(live, "SushiJoin/not-live");
 
         // Verify the queued transaction is targetting one of the dangerous functions on Masterchef
         bytes memory callData;
+        bytes memory argData;
         if (bytes(signature).length == 0) {
             callData = data;
+            argData = new bytes(data.length - 4);
```
Comment (brianmcmichael) [inline src/sushi.sol:187]: Aside from an expensive failed tx, what happens if `data.length < 4` here?

File: src/sushi.sol [hunk 18]
```diff
@@ -75,8 +82,9 @@ contract SushiJoin is CropJoin {
         @param bonus_               The SUSHI token contract address.
         @param masterchef_          The SushiSwap MCV1 contract address.
```
Comment (brianmcmichael) [inline src/sushi.sol:83]:
```suggestion
        @param masterchef_          The SushiSwap MCV2 contract address.
```

**Review by brianmcmichael — CHANGES_REQUESTED** [review]

File: src/sushi.sol (hunk 12 above)
Comment (hexonaut) [reply src/sushi.sol:96]: I mean why not do both? I like having a bunch of sanity checks in case something changes especially with code this heavily dependant on third-party dependencies. Constructor gas is a one-time thing, so I don't see the harm in doing some extra checks for safety.

File: src/sushi.sol (hunk 15 above)
Comment (hexonaut) [reply src/sushi.sol:128]: My intuition is that we use `address` when there is no interface dependency, and we use an interface when we do. In this contract we don't really care about the interface for `migrator` and `rewarder`, we only care about what address they are set to and if that changes. I believe this is consistent with what I've seen elsewhere in dss.

File: src/sushi.sol (hunk 16 above)
Comment (hexonaut) [reply src/sushi.sol:141]: Sure, I'll see if I can come up with something more explicit. This error actually happens quite frequently in the tests, but due to this being an integration test it depends on the state of the MasterChef.

File: src/sushi.sol (hunk 17 above)
Comment (hexonaut) [reply src/sushi.sol:187]: The failure cases of this function are not super important. We only care about if a valid, dangerous transaction [here](https://github.com/sushiswap/sushiswap/blob/master/contracts/governance/Timelock.sol#L115) somehow fails our `cage()`. If the call data is not of the proper abi form in both the signature prepended or standalone variant then [this line](https://github.com/sushiswap/sushiswap/blob/master/contracts/governance/Timelock.sol#L123) will always fail. This is true for any input where `len(signature) == 0` and `len(data) < 4` as there is no function signature.

File: src/sushi.sol (hunk 15 above)
Comment (kmbarry1) [reply src/sushi.sol:128]: For me it's largely a judgment call based on readability and convenience. If something is always used as an interface, sometimes it makes sense to type it as such to cut down on the casting.
